### PR TITLE
Multi agent debate

### PR DESCRIPTION
**Pull request by dylburger**
### <samp>🤖 Generated by Copilot at 3b6aa53</samp>

This pull request improves the code quality, readability, and configurability of the `component_code_gen` package, and adds a new feature that allows generating better component code by trying multiple times and combining the best parts. The main changes are:

### <samp>🤖 Generated by Copilot at 3b6aa53</samp>

> _Sing, O Muse, of the skillful code generation_
> _That the wise and learned LangChain team devised_
> _To create and review components with ease and speed_
> _And to avoid the errors that often arise._

### <samp>🤖 Generated by Copilot at 3b6aa53</samp>

*  Add a new parameter `tries` to control the number of times the model is called to generate code and combine the best parts of each attempt ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L63-R81), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L71-R91), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-38d985e2c8bd73ed773e44c2b976a924b869118a50bef45eabdbe864dff02064L17-R17), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-38d985e2c8bd73ed773e44c2b976a924b869118a50bef45eabdbe864dff02064L24-R29), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-38d985e2c8bd73ed773e44c2b976a924b869118a50bef45eabdbe864dff02064L34-R43), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-38d985e2c8bd73ed773e44c2b976a924b869118a50bef45eabdbe864dff02064L43-R49))
*  Make the temperature parameter configurable and adjustable for different scenarios by using the `OPENAI_TEMPERATURE` environment variable or a default value of 0.5 ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-75d5a3dabdf1cce5000f903be0a7a4b5075a7c679f01e23adf164c8486e7167eR25), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L71-R74))
*  Improve the readability and consistency of the code by following the PEP 8 style guide for imports, comments, and line length ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L1-R74), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L1-L10), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059R6-R14), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L58-R57), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L86-R87))
*  Fix a syntax error and improve the formatting of the `no_docs_system_instructions` variable in the `generate_actions.py` file by adding code block symbols ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fR421), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fL425-R439))
*  Make the `ENABLE_DOCS` environment variable more robust by using a default value of False ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-75d5a3dabdf1cce5000f903be0a7a4b5075a7c679f01e23adf164c8486e7167eL47-R48))
*  Add and modify the instructions and examples for the model and the user in the `no_docs_system_instructions` variable in the `generate_actions.py` file to make them more clear and explicit ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fL4-R11), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fR411-R414), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fL441-R464), [link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-bb92b1907162a142ece31a9550fb28dd055e466a55e7b034c7c464756743ff4fR483-R501))
*  Ignore files with the `.mjs` extension to avoid conflicts with other tools that may not support ES modules ([link](https://github.com/PipedreamHQ/pipedream/pull/7787/files?diff=unified&w=0#diff-61e8bca46b9f9073cf02f150fe9a8860939c84f23ffc996a19ba82cb55c4a2ebR8))

### CODE REVIEW

File: packages/component_code_gen/helpers/langchain_helpers.py [hunk 1]
```diff
@@ -1,18 +1,17 @@
...
-from langchain.agents import ZeroShotAgent, AgentExecutor
-from langchain.chat_models import ChatOpenAI, AzureChatOpenAI
-from langchain.agents.agent_toolkits.json.toolkit import JsonToolkit
-from langchain.tools.json.tool import JsonSpec
 from langchain.schema import (
     # AIMessage,
     HumanMessage,
     SystemMessage
 )
+from langchain.tools.json.tool import JsonSpec
```
Comment (dylburger) [inline packages/component_code_gen/helpers/langchain_helpers.py:6]: These changes were all made by `autopep8`

File: packages/component_code_gen/helpers/langchain_helpers.py [hunk 2]
```diff
@@ -68,10 +67,11 @@ def get_llm():
     if config['openai_api_type'] == "azure":
         azure_config = config["azure"]
         llm = AzureChatOpenAI(deployment_name=azure_config['deployment_name'],
-                model_name=azure_config["model"], temperature=0, request_timeout=300)
+                              model_name=azure_config["model"], temperature=config["temperature"], request_timeout=300)
```
Comment (dylburger) [inline packages/component_code_gen/helpers/langchain_helpers.py:70]: I added this so I could experiment with temperature. Now each agent will generate slightly more creative responses, and the final judge agent can combine and correct.

File: packages/component_code_gen/main.py [hunk 3]
```diff
@@ -14,31 +14,37 @@
...
     return result
 
 
 if __name__ == '__main__':
     parser = argparse.ArgumentParser()
-    parser.add_argument('--type', help='which kind of code you want to generate?', choices=available_templates.keys(), required=True)
-    parser.add_argument('--app', help='the app_name_slug', required=True)
-    parser.add_argument('--instructions', help='markdown file with instructions: prompt + api docs', required=True)
-    parser.add_argument('--verbose', dest='verbose', help='set the logging to debug', required=False, default=False, action='store_true')
+    parser.add_argument('--type', help='Which kind of code you want to generate?',
```
Comment (dylburger) [inline packages/component_code_gen/main.py:35]: Some of these were just formatting changes from `autopep8`

File: packages/component_code_gen/main.py [hunk 4]
```diff
@@ -14,31 +14,37 @@
...
+                        choices=available_templates.keys(), required=True)
+    parser.add_argument('--app', help='The app_name_slug', required=True)
+    parser.add_argument(
+        '--instructions', help='Markdown file with instructions: prompt + api docs', required=True)
+    parser.add_argument('--num_tries', dest='tries', help='The number of times we call the model to generate code',
```
Comment (dylburger) [inline packages/component_code_gen/main.py:40]: The new param for the number of independent agents to run

File: packages/component_code_gen/code_gen/generate_component_code.py [hunk 5]
```diff
@@ -1,76 +1,94 @@
...
-import helpers.supabase_helpers as supabase_helpers
 
-from config.config import config
-import config.logging_config as logging_config
 logger = logging_config.getLogger(__name__)
 
 
-def generate_code(app, prompt, templates):
-    validate_inputs(app, prompt, templates)
+def generate_code(app, prompt, templates, tries):
```
Comment (dylburger) [inline packages/component_code_gen/code_gen/generate_component_code.py:12]: The core change tries to implement [multi-agent debate](https://composable-models.github.io/llm_debate/). The idea is that we get GPT to generate multiple components independently. We raise the temperate to get it to generate more unique output each time. Then we combine those results and make a final call to GPT.

In the code:

* `generate_code` accepts `tries`, an integer.
* We generate code with temperature = 0.5 (default but can be tweaked) `tries` times to get a few different components
* We combine those attempts into a final prompt that asks GPT to judge the code against the rules, picking the best pieces of each, and putting together a final answer.

File: packages/component_code_gen/code_gen/generate_component_code.py [hunk 6]
```diff
@@ -1,76 +1,94 @@
...
+    for idx, result in enumerate(results, 1):
+        new_prompt += f"Try {idx}:\n\n${result}\n\n---\n\n"
+    logger.debug(f"\n\n---------------{new_prompt}\n\n")
+    new_prompt += "---\n\n" + prompt
 
+    # Call the model again with the new prompt to get the final result
+    logger.debug(f"Calling the model a final time to summarize the attempts")
+    return call_langchain(app, new_prompt, templates)
 
-def no_docs(app, prompt, templates):
```
Comment (dylburger) [inline packages/component_code_gen/code_gen/generate_component_code.py:49]: In this refactor I combined `no_docs` and `with_docs` into a single function — `call_langchain`

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**